File: wizards/com/sun/star/wizards/web/data/CGDocument.py
#
# This file is part of the LibreOffice project.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# This file incorporates work covered by the following license notice:
#
#   Licensed to the Apache Software Foundation (ASF) under one or more
#   contributor license agreements. See the NOTICE file distributed
#   with this work for additional information regarding copyright
#   ownership. The ASF licenses this file to you under the Apache
#   License, Version 2.0 (the "License"); you may not use this file
#   except in compliance with the License. You may obtain a copy of
#   the License at http://www.apache.org/licenses/LICENSE-2.0 .
#
from os import sep as separator
from ...common.ConfigGroup import ConfigGroup
from ...document.OfficeDocument import OfficeDocument
from ...common.Properties import Properties
from ...common.PropertyNames import PropertyNames
from ...common.FileAccess import FileAccess
from ...common.XMLHelper import XMLHelper
from ..TypeDetection import *
from ...common.Desktop import Desktop

from com.sun.star.document.MacroExecMode import NEVER_EXECUTE
from com.sun.star.document.UpdateDocMode import NO_UPDATE

'''
About the member fields Title, Description and Author:
There is an ambiguity in handling these fields.
On the one hand, the user can set those fields to a custom value,
relevant only for the web-wizard.
On the other-hand, the web-wizard can read those fields automatically
from SO documents.
In order to avoid conflict by saving, I use two fields: "cp_title", which
should be read from the configuration (saved session), and "title" which is
the value read from the document.
The *cp_title* "overrides" *title*, if exists. if not, *title* is used.
The same is valid for *description* and *author*.
@author  rpiterman
'''

class CGDocument(ConfigGroup):

    PAGE_TYPE_PAGE = 1
    PAGE_TYPE_SLIDE = 2

    def __init__(self, url = "", xmsf = None, task = None):
        self.cp_URL = ""
        self.cp_Exporter = None
        self.cp_Index = -1
        self.cp_Title = ""
        self.cp_Description = ""
        self.cp_Author = ""
        self.localFilename = ""
        self.urlFilename = ""
        self.title = ""
        self.description = ""
        self.author = ""
        self.dirName = ""
        self.createdDate = None
        self.updatedDate = None
        self.sizeBytes = -1
        self.pages = -1
        self.valid = False
        self.appType = ""
        if (xmsf is None):
            return
        self.cp_URL = self.getSettings().getFileAccess(xmsf).getURL(url)
        if (task is None):
            task = Task("", "", 5)
        self.validate(xmsf, task)

    def getSettings(self):
        return self.root

    '''
    the task will advance 5 times during validate.
    @param xmsf
    @param task
    @throws Exception
    '''

    def validate(self, xmsf, task):
        if not self.getSettings().getFileAccess(xmsf).exists(self.cp_URL, False):
            raise FileNotFoundException (
                "The given URL does not point to a file")

        if self.getSettings().getFileAccess(xmsf).isDirectory(self.cp_URL):
            raise IllegalArgumentException ("The given URL points to a directory") #create a TypeDetection service

        self.mediaDescriptor = OfficeDocument.getFileMediaDecriptor(
            xmsf, self.cp_URL)
        task.advance(True)
        #1
        self.analyzeFileType(self.mediaDescriptor)
        task.advance(True)
        #2
        path = self.getSettings().getFileAccess(xmsf).getPath(self.cp_URL, "")
        self.localFilename = FileAccess.getFilename(path, separator)
        '''
        if the type is a star office convertible document
        We try to open the document to get some properties
        '''
        xProps = None
        task.advance(True)
        #3
        if self.isSOOpenable:
            # for documents which are openable through SO,
            # use DocumentProperties service.
            desktop = Desktop.getDesktop(xmsf)
            props = []
            props.append(Properties.createProperty("Hidden", True))
            props.append(Properties.createProperty(
                "MacroExecutionMode", NEVER_EXECUTE))
            props.append(Properties.createProperty(
                "UpdateDocMode", NO_UPDATE))
            component = desktop.loadComponentFromURL(
                self.cp_URL, "_default", 0, tuple(props))
            xProps = component.DocumentProperties

        task.advance(True)
        #4
        #now use the object to read some document properties.
        if xProps is not None:
            self.title = xProps.Title
            self.description = xProps.Description
            self.author = xProps.Author
            self.createdDate = xProps.CreationDate
            self.updatedDate = xProps.ModificationDate
        else:

            #get some information from OS.
            self.title = self.localFilename
            self.updatedDate = \
                self.getSettings().getFileAccess(xmsf).getLastModified(self.cp_URL)

        task.advance(True)
        #5
        if self.cp_Title == "":
            self.cp_Title = self.title

        if self.cp_Title == "":
            self.cp_Title = self.localFilename

        if self.cp_Description == "":
            self.cp_Description = self.description

        if self.cp_Author == "":
            self.cp_Author = self.author

        if self.cp_Exporter is None or self.cp_Exporter == "":
            exp = self.getSettings().getExporters(self.appType)
            self.cp_Exporter = \
                self.getSettings().cp_Exporters.getKey(exp[0])

    '''
    Analyzes a type-detection string, returned from the TypeDetection service,
    and sets the appType, isSOOpenable and isSODocument members.
    '''

    def analyzeFileType(self, mediaDesc):
        if mediaDesc is None:
            media = ""
        else:
            media = Properties.getPropertyValue(
                self.mediaDescriptor, PropertyNames.PROPERTY_NAME)
        self.appType = self.getDocType(media)
        self.isSOOpenable = (self.appType == WRITER_DOC or self.appType == CALC_DOC or self.appType == IMPRESS_DOC or self.appType == DRAW_DOC) or self.appType == HTML_DOC
        parts = media.split("_")
        if len(parts) < 2:
            self.isSODocument = False
        else:
            self.isSODocument = self.isSOOpenable and (parts[1].startswith("Star"))

    '''
    @param media is the media description string returned by an UNO TypeDetection object.
    @return one of the constants in the interface

    '''

    def getDocType(self, media):
        if media == "":
            return NO_TYPE
        elif media.startswith("generic_HTML"):
            return HTML_DOC
        elif media.startswith("writer"):
            return WRITER_DOC
        elif media.startswith("calc"):
            return CALC_DOC
        elif media.startswith("draw"):
            return DRAW_DOC
        elif media.startswith("impress"):
            return IMPRESS_DOC
        elif media.startswith("pdf"):
            return PDF_DOC
        elif media.startswith("gif") or media.startswith("jpg"):
            return WEB_GRAPHICS
        elif media.startswith("wav"):
            return SOUND_FILE
        else:
            return NO_TYPE

    def createDOM(self, parent):
        d = self.getSettings().cp_DefaultSession.cp_Design
        exp = self.getSettings().cp_Exporters.getElement(self.cp_Exporter)
        return XMLHelper.addElement(parent, "document",
                                    ["title", "description", "author", "format", "filename",
                                     "create-date", "update-date", "pages", "size", "icon",
                                     "dir", "fn"],
                                    [self.cp_Title if (d.cp_DisplayTitle) else "",
                                     self.cp_Description if (d.cp_DisplayDescription) else "",
                                     self.cp_Author if (d.cp_DisplayAuthor) else "",
                                     self.getTargetTypeName(exp) if (d.cp_DisplayFileFormat) else "",
                                     self.localFilename if (d.cp_DisplayFilename) else "",
                                     self.createDate() if (d.cp_DisplayCreateDate) else "",
                                     self.updateDate() if (d.cp_DisplayUpdateDate) else  "",
                                     "" + self.getPages() if (d.cp_DisplayPages and (self.pages > -1)) else "", #TODO when do i calculate pages?
                                     self.sizeKB() if (d.cp_DisplaySize) else "", #TODO when do i calculate size?
                                     self.getIcon(exp) if (d.cp_DisplayFormatIcon) else "",
                                     self.dirName, self.urlFilename])

    def updateDate(self):
        if self.updatedDate is None:
            return ""

        return self.getSettings().formatter.formatCreated(self.updatedDate)

    def createDate(self):
        if self.createdDate is None:
            return ""

        return self.getSettings().formatter.formatCreated(self.createdDate)

    def sizeKB(self):
        if self.sizeBytes == -1:
            return ""
        else:
            return self.getSettings().formatter.formatFileSize(self.sizeBytes)

    def getPages(self):
        if self.pages == -1:
            return ""
        else:
            return self.pagesTemplate().replace("%NUMBER", "" + self.pages)

    def pagesTemplate(self):
        pagesType = \
            self.getSettings().cp_Exporters.getElement(self.cp_Exporter).cp_PageType
        if pagesType == PAGE_TYPE_PAGE:
            return \
                self.getSettings().resources[CGSettings.RESOURCE_PAGES_TEMPLATE]
        elif pagesType == PAGE_TYPE_SLIDE:
            return \
                self.getSettings().resources[CGSettings.RESOURCE_SLIDES_TEMPLATE]
        else:
            return ""

    def getTargetTypeName(self, exp):
        if exp.targetTypeName == "":
            return Properties.getPropertyValue(self.mediaDescriptor, "UIName")
        else:
            return exp.targetTypeName

    def getIcon(self, exporter):
        if exporter.cp_Icon == "":
            return self.getIcon1(self.appType)
        else:
            return exporter.cp_Icon

    def getIcon1(self, appType):
        return appType + ".gif"

    '''
    This method is used by the DataAware model
    and returns the index of the exporter of
    this document, <b>in the list of the
    exporters available for this type of document</b>
    @return
    '''

    def getExporter(self):
        return [self.getExporterIndex()]

    '''
    @see getExporter()
    '''

    def setExporter(self, exporter_):
        exp = self.getSettings().getExporters(self.appType)[exporter_[0]]
        self.cp_Exporter = self.getSettings().cp_Exporters.getKey(exp)

    '''
    @see getExporter()
    @return
    '''

    def getExporterIndex(self):
        if self.cp_Exporter is None:
            return 0

        exporter = self.getSettings().cp_Exporters.getElement(self.cp_Exporter)
        exporters = self.getSettings().getExporters(self.appType)
        i = 0
        while i < len(exporters):
            if exporters[i] == exporter:
                return i

            i += 1
        return -1

    def toString(self):
        return self.localFilename

File: wizards/com/sun/star/wizards/web/data/CGDocument.java
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */
package com.sun.star.wizards.web.data;

import java.io.File;
import java.io.FileNotFoundException;

import org.w3c.dom.Node;

import com.sun.star.beans.PropertyValue;
import com.sun.star.document.MacroExecMode;
import com.sun.star.document.UpdateDocMode;
import com.sun.star.document.XDocumentProperties;
import com.sun.star.document.XDocumentPropertiesSupplier;
import com.sun.star.frame.XComponentLoader;
import com.sun.star.frame.XDesktop;
import com.sun.star.lang.XComponent;
import com.sun.star.lang.XMultiServiceFactory;
import com.sun.star.uno.UnoRuntime;
import com.sun.star.util.DateTime;
import com.sun.star.wizards.common.Desktop;
import com.sun.star.wizards.common.FileAccess;
import com.sun.star.wizards.common.JavaTools;
import com.sun.star.wizards.common.Properties;
import com.sun.star.wizards.common.PropertyNames;
import com.sun.star.wizards.common.XMLHelper;
import com.sun.star.wizards.common.XMLProvider;
import com.sun.star.wizards.document.OfficeDocument;
import com.sun.star.wizards.ui.event.Task;

/**
 * About the member fields Title, Decription and Author:
 * There is an ambiguicy in handling these fields.
 * On the one hand, the user can set those fields to a custom value,
 * relevant only for the web-wizard.
 * On the other-hand, the web-wizard can read those fields automatically
 * from SO documents.
 * In order to avoid conflict by saving, I use two fields: "cp_title", which
 * should be read from the configuratin (saved session), and "title" which is
 * the value read from the document.
 * The *cp_title* "overrides" *title*, if exists. if not, *title* is used.
 * The same is valid for *description* and *author*.
 */
public class CGDocument extends ConfigSetItem implements XMLProvider
{

    public static final int PAGE_TYPE_UNKNOWN = 0;
    public static final int PAGE_TYPE_PAGE = 1;
    public static final int PAGE_TYPE_SLIDE = 2;
    public String cp_Title = PropertyNames.EMPTY_STRING;
    public String cp_Description = PropertyNames.EMPTY_STRING;
    public String cp_URL;
    public String cp_Author = PropertyNames.EMPTY_STRING;
    public String cp_Exporter;
    /**
     * the destination filename to which this
     * document will be exported to.
     */
    public String localFilename = PropertyNames.EMPTY_STRING;
    public String urlFilename = PropertyNames.EMPTY_STRING;
    public String title = PropertyNames.EMPTY_STRING;
    public String description = PropertyNames.EMPTY_STRING;
    public String author = PropertyNames.EMPTY_STRING;
    public DateTime createDate;
    public DateTime updateDate;
    public int sizeBytes = -1;
    public int pages = -1;
    /**
     * The contents subdirectory name in which the document is in.
     * This subdirectory will be created in the content's "docs"
     * subdirectory.  
     */
    public String dirName;
    /**
     * contains an identifier of the type of this document.
     */
    public String appType;
    public PropertyValue[] mediaDescriptor;
    /**
     * Is this document an original SO Document (different versions).
     */
    public boolean isSODocument;
    /**
     * can this document be opend through SO?
     */
    public boolean isSOOpenable;
    /**
     * This flag is set to true after the validate method is called.
     */
    public boolean valid = false;
    public String targetTypeName;

    public CGDocument()
    {
    }

    public CGDocument(String url, XMultiServiceFactory xmsf, Task task) throws Exception
    {
        cp_URL = getSettings().getFileAccess(xmsf).getURL(url);
        if (task == null)
        {
            task = new Task(PropertyNames.EMPTY_STRING, PropertyNames.EMPTY_STRING, 5);
        }
        validate(xmsf, task);
    }

    /**
     * the task will advance 5 times during validate.
     * @param xmsf
     * @param task
     * @throws Exception
     */
    public void validate(XMultiServiceFactory xmsf, Task task)
            throws FileNotFoundException,
            IllegalArgumentException,
            Exception
    {

        if (!getSettings().getFileAccess(xmsf).exists(cp_URL, false))
        {
            throw new FileNotFoundException("The given URL does not point to a file");
        }
        if (getSettings().getFileAccess(xmsf).isDirectory(cp_URL))
        {
            throw new IllegalArgumentException("The given URL points to a directory");        //create a TypeDetection service
        }
        mediaDescriptor = OfficeDocument.getFileMediaDecriptor(xmsf, cp_URL);

        task.advance(true); //1

        analyzeFileType(mediaDescriptor);

        task.advance(true); //2

        String path = getSettings().getFileAccess(xmsf).getPath(cp_URL, PropertyNames.EMPTY_STRING);
        localFilename = FileAccess.getFilename(path, File.separator);

        /* if the type is a star office convertable document
         * We try to open the document to get some properties
         */

        XDocumentProperties xProps = null;

        task.advance(true); //3

        if (isSOOpenable)
        { // for documents which are openable through SO, use DocumentProperties service.
            XDesktop desktop = Desktop.getDesktop(xmsf);
            PropertyValue[] props = new PropertyValue[3];
            props[0] = Properties.createProperty("Hidden", Boolean.TRUE);
            props[1] = Properties.createProperty("MacroExecutionMode", new Short(MacroExecMode.NEVER_EXECUTE));
            props[2] = Properties.createProperty("UpdateDocMode", new Short(UpdateDocMode.NO_UPDATE));
            XComponent component = UnoRuntime.queryInterface(XComponentLoader.class, desktop).loadComponentFromURL(cp_URL, "_default", 0, props);
            xProps = UnoRuntime.queryInterface(XDocumentPropertiesSupplier.class, component).getDocumentProperties();
        }

        task.advance(true); //4    

        //now use the object to read some document properties.
        if (xProps != null)
        {
            title = xProps.getTitle();
            description = xProps.getDescription();
            author = xProps.getAuthor();
            createDate = xProps.getCreationDate();
            updateDate = xProps.getModificationDate();
        }
        else
        { //get some information from OS.
            title = localFilename;
            updateDate = getSettings().getFileAccess(xmsf).getLastModified(cp_URL);
        }

        task.advance(true); //5

        valid = true;

        if (cp_Title.equals(PropertyNames.EMPTY_STRING))
        {
            cp_Title = title;
        }
        if (cp_Title.equals(PropertyNames.EMPTY_STRING))
        {
            cp_Title = localFilename;
        }
        if (cp_Description.equals(PropertyNames.EMPTY_STRING))
        {
            cp_Description = description;
        }
        if (cp_Author.equals(PropertyNames.EMPTY_STRING))
        {
            cp_Author = author;
        }
        if (cp_Exporter == null || cp_Exporter.equals(PropertyNames.EMPTY_STRING))
        {
            cp_Exporter = (String) getSettings().cp_Exporters.getKey(
                    getSettings().getExporters(appType)[0]);
        }
    }

    /**
     * Analyzes a type-detection string, returned from the TypeDetection service,
     * and sets the appType, isSOOpenable and isSODocument members.
     */
    private void analyzeFileType(PropertyValue[] mediaDesc)
    {

        if (mediaDesc == null)
        {
        }

        String media = (mediaDesc == null)
                ? PropertyNames.EMPTY_STRING : (String) Properties.getPropertyValue(mediaDescriptor, PropertyNames.PROPERTY_NAME);
        appType = getDocType(media);

        isSOOpenable =
                (appType.equals(TypeDetection.WRITER_DOC) || appType.equals(TypeDetection.CALC_DOC) || appType.equals(TypeDetection.IMPRESS_DOC) || appType.equals(TypeDetection.DRAW_DOC)) || appType.equals(TypeDetection.HTML_DOC);

        String[] parts = JavaTools.ArrayoutofString(media, "_");


        isSODocument = parts.length >= 2 && isSOOpenable && (parts[1].startsWith("Star"));

    }

    /**
     * @param media is the media description string returned by an UNO TypeDetection object.
     * @return one of the constants in the interface TypeDetection.
     * 
     */
    private String getDocType(String media)
    {
        if (media.equals(PropertyNames.EMPTY_STRING))
        {
            return TypeDetection.NO_TYPE;
        }
        if (media.startsWith("generic_HTML"))
        {
            return TypeDetection.HTML_DOC;
        }
        if (media.startsWith("writer"))
        {
            return TypeDetection.WRITER_DOC;
        }
        else if (media.startsWith("calc"))
        {
            return TypeDetection.CALC_DOC;
        }
        else if (media.startsWith("draw"))
        {
            return TypeDetection.DRAW_DOC;
        }
        else if (media.startsWith("impress"))
        {
            return TypeDetection.IMPRESS_DOC;
        }
        else if (media.startsWith("pdf"))
        {
            return TypeDetection.PDF_DOC;
        }
        else if (media.startsWith("gif") || media.startsWith("jpg"))
        {
            return TypeDetection.WEB_GRAPHICS;
        }
        else if (media.startsWith("wav"))
        {
            return TypeDetection.SOUND_FILE;
        }
        else
        {
            return TypeDetection.NO_TYPE;
        }
    }

    public Node createDOM(Node parent)
    {
        CGDesign d = getSettings().cp_DefaultSession.cp_Design;
        CGExporter exp = (CGExporter) getSettings().cp_Exporters.getElement(cp_Exporter);

        return XMLHelper.addElement(parent, "document",
                new String[]
                {
                    "title", "description", "author", "format", "filename", "create-date", "update-date", "pages", "size", "icon", "dir", "fn"
                },
                new String[]
                {
                    d.cp_DisplayTitle ? cp_Title : PropertyNames.EMPTY_STRING,
                    d.cp_DisplayDescription ? cp_Description : PropertyNames.EMPTY_STRING,
                    d.cp_DisplayAuthor ? cp_Author : PropertyNames.EMPTY_STRING,
                    d.cp_DisplayFileFormat ? getTargetTypeName(exp) : PropertyNames.EMPTY_STRING,
                    d.cp_DisplayFilename ? localFilename : PropertyNames.EMPTY_STRING,
                    d.cp_DisplayCreateDate ? createDate() : PropertyNames.EMPTY_STRING,
                    d.cp_DisplayUpdateDate ? updateDate() : PropertyNames.EMPTY_STRING,
                    d.cp_DisplayPages && (pages > -1) ? PropertyNames.EMPTY_STRING + pages() : PropertyNames.EMPTY_STRING, //TODO when do i calculate pages?
                    d.cp_DisplaySize ? sizeKB() : PropertyNames.EMPTY_STRING,//TODO when do i calculate size?
                    d.cp_DisplayFormatIcon ? getIcon(exp) : PropertyNames.EMPTY_STRING,
                    dirName, urlFilename
                });
    }

    private String updateDate()
    {
        if (this.updateDate == null)
        {
            return PropertyNames.EMPTY_STRING;
        }
        return getSettings().formatter.formatCreated(this.updateDate);
    }

    private String createDate()
    {
        if (this.createDate == null)
        {
            return PropertyNames.EMPTY_STRING;
        }
        return getSettings().formatter.formatCreated(this.createDate);
    }

    private String sizeKB()
    {
        if (sizeBytes == -1)
        {
            return PropertyNames.EMPTY_STRING;
        }
        else
        {
            return getSettings().formatter.formatFileSize(sizeBytes);
        }
    }

    private String pages()
    {
        return pages == -1 ? PropertyNames.EMPTY_STRING
                : JavaTools.replaceSubString(pagesTemplate(), PropertyNames.EMPTY_STRING + pages, "%NUMBER");
    }

    private String pagesTemplate()
    {
        int pagesType = ((CGExporter) getSettings().cp_Exporters.getElement(cp_Exporter)).cp_PageType;
        switch (pagesType)
        {
            case PAGE_TYPE_UNKNOWN:
                return PropertyNames.EMPTY_STRING;
            case PAGE_TYPE_PAGE:
                return getSettings().resources[CGSettings.RESOURCE_PAGES_TEMPLATE];
            case PAGE_TYPE_SLIDE:
                return getSettings().resources[CGSettings.RESOURCE_SLIDES_TEMPLATE];
            default:
                return PropertyNames.EMPTY_STRING;
        }
    }

    private String getTargetTypeName(CGExporter exp)
    {
        return (exp.targetTypeName.equals(PropertyNames.EMPTY_STRING))
                ? (String) Properties.getPropertyValue(mediaDescriptor, "UIName")
                : exp.targetTypeName;
    }

    private String getIcon(CGExporter exporter)
    {
        return exporter.cp_Icon.equals(PropertyNames.EMPTY_STRING) ? getIcon(this.appType) : exporter.cp_Icon;
    }

    private String getIcon(String appType)
    {
        return appType + ".gif";
    }

    /**
     * This method is used by the DataAware model
     * and returns the index of the exporter of
     * this document, <b>in the list of the
     * exporters available for this type of document</b>
     * @return
     */
    public short[] getExporter()
    {
        return new short[]
                {
                    getExporterIndex()
                };
    }

    /**
     * @see #getExporter()
     */
    public void setExporter(short[] exporter_)
    {
        Object exp = getSettings().getExporters(appType)[exporter_[0]];
        cp_Exporter = (String) getSettings().cp_Exporters.getKey(exp);
    }

    /**
     * @see #getExporter()
     * @return
     */
    private short getExporterIndex()
    {
        if (cp_Exporter == null)
        {
            return 0;
        }
        Object exporter = getSettings().cp_Exporters.getElement(cp_Exporter);
        Object[] exporters = getSettings().getExporters(appType);
        for (short i = 0; i < exporters.length; i++)
        {
            if (exporters[i] == exporter)
            {
                return i;
            }
        }
        return -1;
    }

    public String toString()
    {
        return localFilename;
    }
}
